swarm/network/stream/stream.go

```diff
@@ -509,6 +492,69 @@ func (r *Registry) updateSyncing() {
 	}
 }
 
+// requestPeerSubscriptions calls on each live peer in the kademlia table
+// and sends a `RequestSubscription` to peers according to their bin
+// and their relationship with kademlia's depth.
+// Also check `TestRequestPeerSubscriptions` in order to understand the
+// expected behavior.
+// The function expects:
+//   * the kademlia
+//   * a map of subscriptions
+//   * the actual function to subscribe
+//     (in case of the test, it doesn't do real subscriptions)
+func (r *Registry) requestPeerSubscriptions(
+	kad *network.Kademlia,
+	subs map[enode.ID]map[Stream]struct{},
+	subscriptionFunc func(p *network.Peer, bin uint8, subs map[enode.ID]map[Stream]struct{}) bool) {
+
+	var startPo int
+	var endPo int
+	var ok bool
+
+	// kademlia's depth
+	kadDepth := kad.NeighbourhoodDepth()
+	// request subscriptions for all nodes and bins
+	// nil as base takes the node's base; we need to pass 255 as `EachConn` runs
+	// from deepest bins backwards
+	kad.EachConn(nil, 255, func(p *network.Peer, po int, _ bool) bool {
+		//if the peer's bin is shallower than the kademlia depth,
+		//only the peer's bin should be subscribed
+		if po < kadDepth {
+			startPo = po
+			endPo = po
+		} else {
+			//if the peer's bin is equal or deeper than the kademlia depth,
+			//each bin from the depth up to k.MaxProxDisplay should be subscribed
+			startPo = kadDepth
+			endPo = kad.MaxProxDisplay
+		}
+
+		for bin := startPo; bin <= endPo; bin++ {
+			//do the actual subscription
+			ok = subscriptionFunc(p, uint8(bin), subs)
+		}
+		return ok
+	})
+}
+
+// doRequestSubscription sends the actual RequestSubscription to the peer
+func (r *Registry) doRequestSubscription(p *network.Peer, bin uint8, subs map[enode.ID]map[Stream]struct{}) bool {
+	log.Debug(fmt.Sprintf("Requesting subscription by: registry %s from peer %s for bin: %d", r.addr, p.ID(), bin))
+	// bin is always less then 256 and it is safe to convert it to type uint8
+	stream := NewStream("SYNC", FormatSyncBinKey(uint8(bin)), true)
+	if streams, ok := subs[p.ID()]; ok {
+		// delete live and history streams from the map, so that it won't be removed with a Quit request
```

Comment:
This is code which was existing prior to my edits/changes. 

TBH, I don't know why need to check these `subs` and delete `live` and `history` streams which match. To me it looks **obsolete**. @zelig ?

Reply:
Not addressed. I think we should be clear about what these do.

Reply:
@zelig what's the need for this code? I just copied it over from the existing version

Reply:
i dont understand really either 😊

Reply:
Should we be brave then and remove it?

Reply:
https://gitter.im/ethersphere/orange-lounge?at=5c37c97c6a394e28218a7a2d 

Check this comment by @janos, we may need to keep this as-is

Reply:
If another version of this code would help for better understanding, feel free to improve it. But I remember that we talked about it earlier, with @nolash in Vienna when the first version was written, and I believe with @holisticode when first syncing tests from snapshots were written. This is a clear sign that the function is very badly written and that it needs to be changed.

Reply:
I've read through your discussion in orange lounge @janos and @holisticode and I must admit I am at a loss to understand the details since I don't really know the stream package very well.

It seems your conclusion is that the code is bad, but we leave it for the sake of having some closure of the efforts contained in this PR. I can approve, but please under the condition that the problems with the code are addressed in a new issue with an explanation that even I can understand :)